<a href="https://colab.research.google.com/github/jeffdubo/deep-learning-challenge/blob/main/charity_venture_funding.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Preprocessing

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [2]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df.drop(columns=['EIN', 'NAME'], inplace=True)

In [3]:
# Determine the number of unique values in each column.
application_df.nunique()

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64

In [4]:
# Look at APPLICATION_TYPE value counts for binning
application_type_counts = application_df['APPLICATION_TYPE'].value_counts()
application_type_counts

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [5]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
application_types_to_replace = [i for i in application_type_counts.index if application_type_counts[i] < 500]

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [6]:
# Look at CLASSIFICATION value counts for binning
classification_counts = application_df['CLASSIFICATION'].value_counts()
classification_counts

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [7]:
# You may find it helpful to look at CLASSIFICATION value counts >1
classification_counts[classification_counts > 1]

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: CLASSIFICATION, dtype: int64

In [8]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
# Code Ref: https://stackoverflow.com/questions/16476924/how-to-iterate-over-rows-in-a-dataframe-in-pandas
classifications_to_replace = [i for i in classification_counts.index if classification_counts[i] < 1000]

# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")

# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: CLASSIFICATION, dtype: int64

In [9]:
application_df.head()

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0              T10       Independent          C1000    ProductDev   
1               T3       Independent          C2000  Preservation   
2               T5  CompanySponsored          C3000    ProductDev   
3               T3  CompanySponsored          C2000  Preservation   
4               T3       Independent          C1000     Heathcare   

   ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  \
0   Association       1              0                      N     5000   
1  Co-operative       1         1-9999                      N   108590   
2   Association       1              0                      N     5000   
3         Trust       1    10000-24999                      N     6692   
4         Trust       1  100000-499999                      N   142590   

   IS_SUCCESSFUL  
0              1  
1              1  
2              0  
3              1  
4              1

In [10]:
# Convert categorical data to numeric with `pd.get_dummies`
application_df = pd.get_dummies(application_df)
application_df.head()

STATUS  ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_Other  \
0       1     5000              1                       0   
1       1   108590              1                       0   
2       1     5000              0                       0   
3       1     6692              1                       0   
4       1   142590              1                       0   

   APPLICATION_TYPE_T10  APPLICATION_TYPE_T19  APPLICATION_TYPE_T3  \
0                     1                     0                    0   
1                     0                     0                    1   
2                     0                     0                    0   
3                     0                     0                    1   
4                     0                     0                    1   

   APPLICATION_TYPE_T4  APPLICATION_TYPE_T5  APPLICATION_TYPE_T6  ...  \
0                    0                    0                    0  ...   
1                    0                    0                    0  ...   
2                    0                    1                    0  ...   
3                    0                    0                    0  ...   
4                    0                    0                    0  ...   

   INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0                  0                       0                         0   
1                  1                       0                         0   
2                  0                       0                         0   
3                  0                       1                         0   
4                  0                       0                         1   

   INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                   0                 0                       0   
1                   0                 0                       0   
2                   0                 0                       0   
3                   0                 0                       0   
4                   0                 0                       0   

   INCOME_AMT_50M+  INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  \
0                0                  0                         1   
1                0                  0                         1   
2                0                  0                         1   
3                0                  0                         1   
4                0                  0                         1   

   SPECIAL_CONSIDERATIONS_Y  
0                         0  
1                         0  
2                         0  
3                         0  
4                         0  

[5 rows x 44 columns]

In [13]:
# Split our preprocessed data into our features and target arrays
X = application_df.drop(columns='IS_SUCCESSFUL')
y = application_df['IS_SUCCESSFUL']

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=78)

# Display value counts for y_train and y_test to check distribution of values
print('y_train:')
print(y_train.value_counts())
print('\ny_test:')
print(y_test.value_counts())

y_train:
1    13689
0    12035
Name: IS_SUCCESSFUL, dtype: int64

y_test:
1    4572
0    4003
Name: IS_SUCCESSFUL, dtype: int64


In [14]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [19]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.

# Create function to define model and compile to test different parameters
def create_model(num_input_features, nodes_l1, nodes_l2, nodes_l3, activation_l1, activation, optimizer):

    # Clear old sessions/models
    # Code Ref: https://www.tensorflow.org/api_docs/python/tf/keras/backend/clear_session
    tf.keras.backend.clear_session()
    nn = tf.keras.models.Sequential()

    # Hidden Layers
    nn.add(tf.keras.layers.Dense(units=nodes_l1, input_dim=num_input_features, activation=activation_l1))
    nn.add(tf.keras.layers.Dense(units=nodes_l2, activation=activation))
    # Add 3rd layer if nodes_l3 > 0
    if nodes_l3 > 0:
        nn.add(tf.keras.layers.Dense(units=nodes_l3, activation=activation))
    # Output layer
    nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

    # Check the structure of the model
    display(nn.summary())

    # Compile the model
    nn.compile(loss='binary_crossentropy', optimizer=optimizer, metrics=['accuracy'])

    return nn

In [20]:
# Create and train Model #1
#   First Layer: 80 nodes, relu activation
#   Second Layer: 30 nodes, relu activation
#   Output Layer: adam optimizer
num_input_features = len(X_train.iloc[0])
nn = create_model(num_input_features, 80, 30, 0, 'relu', 'relu', 'adam')
fit_model = nn.fit(X_train_scaled, y_train, epochs=100)

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 80)                3520      
                                                                 
 dense_1 (Dense)             (None, 30)                2430      
                                                                 
 dense_2 (Dense)             (None, 1)                 31        
                                                                 
Total params: 5981 (23.36 KB)
Trainable params: 5981 (23.36 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


None

Epoch 1/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5701 - accuracy: 0.7225
Epoch 2/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5550 - accuracy: 0.7306
Epoch 3/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5528 - accuracy: 0.7311
Epoch 4/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5513 - accuracy: 0.7319
Epoch 5/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5505 - accuracy: 0.7329
Epoch 6/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5489 - accuracy: 0.7332
Epoch 7/100
804/804 [==============================] - 5s 6ms/step - loss: 0.5484 - accuracy: 0.7325
Epoch 8/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5472 - accuracy: 0.7343
Epoch 9/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5470 - accuracy: 0.7350
Epoch 10/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5460 - accura

In [21]:
# Evaluate Model #1
#   First Layer: 80 nodes, relu activation
#   Second Layer: 30 nodes, relu activation
#   Output Layer: adam optimizer
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5556 - accuracy: 0.7249 - 519ms/epoch - 2ms/step
Loss: 0.5556123852729797, Accuracy: 0.7248979806900024


### Analysis: Model #1 did not perform well with an accuracy of only 0.725 and loss of .556.

In [22]:
# Mount Google Drive in Colab
# Code Ref: https://medium.com/@ml_kid/how-to-save-our-model-to-google-drive-and-reuse-it-2c1028058cb2
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [31]:
# Export model #1 to HDF5 file in shared Google folder
path = F'/content/drive/My Drive/Bootcamp/Challenge-Deep-Learning/charity_dl_model_1.H5'
nn.save(path)

## Create, Train and Evaluate Additional Models

In [22]:
# Create and train Model #2
#   First Layer: 80 nodes, relu activation
#   Second Layer: 30 nodes, relu activation
#   Output Layer: Nadam optimizer
num_input_features = len(X_train.iloc[0])
nn = create_model(num_input_features, 80, 30, 0, 'relu', 'relu', 'Nadam')
fit_model = nn.fit(X_train_scaled, y_train, epochs=100)

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 80)                3520      
                                                                 
 dense_1 (Dense)             (None, 30)                2430      
                                                                 
 dense_2 (Dense)             (None, 1)                 31        
                                                                 
Total params: 5981 (23.36 KB)
Trainable params: 5981 (23.36 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


None

Epoch 1/100
804/804 [==============================] - 4s 2ms/step - loss: 0.5716 - accuracy: 0.7218
Epoch 2/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5548 - accuracy: 0.7301
Epoch 3/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5520 - accuracy: 0.7307
Epoch 4/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5503 - accuracy: 0.7323
Epoch 5/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5491 - accuracy: 0.7331
Epoch 6/100
804/804 [==============================] - 5s 6ms/step - loss: 0.5489 - accuracy: 0.7332
Epoch 7/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5473 - accuracy: 0.7339
Epoch 8/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5470 - accuracy: 0.7339
Epoch 9/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5469 - accuracy: 0.7343
Epoch 10/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5461 - accura

In [23]:
# Evaluate Model #2
#   First Layer: 80 nodes, relu activation
#   Second Layer: 30 nodes, relu activation
#   Output Layer: Nadam optimizer
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5574 - accuracy: 0.7272 - 513ms/epoch - 2ms/step
Loss: 0.5573615431785583, Accuracy: 0.7272303104400635


### Analysis: Very slight improvement in accuracy by 0.0023, but an increase in loss by .0002.

In [24]:
# Create and train Model #3
#   First Layer: 100 nodes, relu activation
#   Second Layer: 70 nodes, relu activation
#   Third Layer:  30 nodes, relu activation
#   Output Layer: Nadam optimizer
num_input_features = len(X_train.iloc[0])
nn = create_model(num_input_features, 100, 70, 30, 'relu', 'relu', 'Nadam')
fit_model = nn.fit(X_train_scaled, y_train, epochs=100)

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 100)               4400      
                                                                 
 dense_1 (Dense)             (None, 70)                7070      
                                                                 
 dense_2 (Dense)             (None, 30)                2130      
                                                                 
 dense_3 (Dense)             (None, 1)                 31        
                                                                 
Total params: 13631 (53.25 KB)
Trainable params: 13631 (53.25 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


None

Epoch 1/100
804/804 [==============================] - 7s 6ms/step - loss: 0.5707 - accuracy: 0.7211
Epoch 2/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5557 - accuracy: 0.7292
Epoch 3/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5522 - accuracy: 0.7324
Epoch 4/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5498 - accuracy: 0.7332
Epoch 5/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5495 - accuracy: 0.7339
Epoch 6/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5475 - accuracy: 0.7346
Epoch 7/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5481 - accuracy: 0.7346
Epoch 8/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5467 - accuracy: 0.7352
Epoch 9/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5460 - accuracy: 0.7353
Epoch 10/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5459 - accura

In [25]:
Evaluate Model #3
#   First Layer: 100 nodes, relu activation
#   Second Layer: 70 nodes, relu activation
#   Third Layer:  30 nodes, relu activation
#   Output Layer: Nadam optimizer
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5633 - accuracy: 0.7251 - 532ms/epoch - 2ms/step
Loss: 0.5632742047309875, Accuracy: 0.7251312136650085


### Analysis: There was no signficant improvement in accuracy and loss from the first model and this model underperformed in comparison to the second model.

In [26]:
# Create and train Model #4
#   First Layer: 100 nodes, relu activation
#   Second Layer: 70 nodes, tanh activation
#   Output Layer: Nadam optimizer
num_input_features = len(X_train.iloc[0])
nn = create_model(num_input_features, 80, 30, 0, 'relu', 'tanh', 'Nadam')
fit_model = nn.fit(X_train_scaled, y_train, epochs=100)

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 80)                3520      
                                                                 
 dense_1 (Dense)             (None, 30)                2430      
                                                                 
 dense_2 (Dense)             (None, 1)                 31        
                                                                 
Total params: 5981 (23.36 KB)
Trainable params: 5981 (23.36 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


None

Epoch 1/100
804/804 [==============================] - 3s 2ms/step - loss: 0.5692 - accuracy: 0.7206
Epoch 2/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5563 - accuracy: 0.7286
Epoch 3/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5527 - accuracy: 0.7313
Epoch 4/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5511 - accuracy: 0.7330
Epoch 5/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5504 - accuracy: 0.7322
Epoch 6/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5488 - accuracy: 0.7327
Epoch 7/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5487 - accuracy: 0.7326
Epoch 8/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5478 - accuracy: 0.7338
Epoch 9/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5476 - accuracy: 0.7338
Epoch 10/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5470 - accura

In [27]:
# Evaluate Model #4
#   First Layer: 100 nodes, relu activation
#   Second Layer: 70 nodes, tanh activation
#   Output Layer: Nadam optimizer
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5528 - accuracy: 0.7244 - 552ms/epoch - 2ms/step
Loss: 0.5527868866920471, Accuracy: 0.7244315147399902


### Analysis: No signification improvement. Accuracy increased by 0.0001 from first model with the lowest loss of all models.

## Create, Train, and Evaluate Additional Models using Keras Tuner

In [31]:
# Create a function that creates a new Sequential model with hyperparameter options
def create_model_hp(hp):

    tf.keras.backend.clear_session()
    nn = tf.keras.models.Sequential()

    # Allow kerastuner to decide which activation function to use in hidden layers
    activation = hp.Choice('activation',['relu','tanh'])

    # Allow kerastuner to decide number of neurons in first layer
    nn.add(tf.keras.layers.Dense(units=hp.Int('first_units',
        min_value=50,
        max_value=100,
        step=10), activation=activation, input_dim=num_input_features))

    # Allow kerastuner to decide number of hidden layers and neurons in hidden layers
    for i in range(hp.Int('num_layers', 1, 5)):
        nn.add(tf.keras.layers.Dense(units=hp.Int('units_' + str(i),
            min_value=10,
            max_value=50,
            step=10),
            activation=activation))

    nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

    # Compile the model
    nn.compile(loss='binary_crossentropy', optimizer='Nadam', metrics=['accuracy'])

    return nn

In [29]:
!pip install keras_tuner

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 129.5/129.5 kB 2.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 950.8/950.8 kB 26.3 MB/s eta 0:00:00


In [ ]:
# Import the kerastuner library
import keras_tuner as kt

# Initiate tuner
tuner = kt.Hyperband(
    create_model_hp,
    objective='val_accuracy',
    max_epochs=30,
    hyperband_iterations=2)

tuner.search(X_train_scaled, y_train, epochs=20, validation_data=(X_test_scaled, y_test))

Trial 159 Complete [00h 00m 24s]
val_accuracy: 0.7287463545799255

Best val_accuracy So Far: 0.7294460535049438
Total elapsed time: 01h 03m 01s

Search: Running Trial #160

Value             |Best Value So Far |Hyperparameter
relu              |relu              |activation
100               |70                |first_units
2                 |3                 |num_layers
10                |30                |units_0
50                |30                |units_1
40                |20                |units_2
40                |10                |units_3
20                |20                |units_4
10                |10                |tuner/epochs
4                 |0                 |tuner/initial_epoch
2                 |1                 |tuner/bracket
1                 |0                 |tuner/round
0155              |None              |tuner/trial_id

Epoch 5/10
804/804 [==============================] - 5s 4ms/step - loss: 0.5496 - accuracy: 0.7328 - val_loss: 0.5578 - val_accura

In [36]:
# Display evaluation and hyperparamters for top 3 models
for i in list(range(0, 3)):
    # Get model and hyperparameters
    model = tuner.get_best_models(3)[i]
    hyper = tuner.get_best_hyperparameters(3)[i]

    # Evaluate model
    model_loss, model_accuracy = model.evaluate(X_test_scaled,y_test,verbose=2)

    # Print model evaluation and hyperparamaters
    print(f'Model Rated #{i+1} Evaluation:')
    print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")
    print(f'Model Rated #{i+1} Hyperparamaters:')
    display(hyper.values)

268/268 - 1s - loss: 0.5495 - accuracy: 0.7294 - 825ms/epoch - 3ms/step
Model Rated #1 Evaluation:
Loss: 0.5494797229766846, Accuracy: 0.7294460535049438
Model Rated #1 Hyperparamaters:


{'activation': 'relu',
 'first_units': 70,
 'num_layers': 4,
 'units_0': 20,
 'units_1': 20,
 'units_2': 10,
 'units_3': 10,
 'units_4': 10,
 'tuner/epochs': 20,
 'tuner/initial_epoch': 7,
 'tuner/bracket': 1,
 'tuner/round': 1,
 'tuner/trial_id': '0021'}

268/268 - 1s - loss: 0.5504 - accuracy: 0.7292 - 691ms/epoch - 3ms/step
Model Rated #2 Evaluation:
Loss: 0.5503792762756348, Accuracy: 0.7292128205299377
Model Rated #2 Hyperparamaters:


{'activation': 'relu',
 'first_units': 60,
 'num_layers': 3,
 'units_0': 10,
 'units_1': 30,
 'units_2': 30,
 'units_3': 20,
 'units_4': 20,
 'tuner/epochs': 20,
 'tuner/initial_epoch': 7,
 'tuner/bracket': 2,
 'tuner/round': 2,
 'tuner/trial_id': '0045'}

268/268 - 1s - loss: 0.5506 - accuracy: 0.7292 - 757ms/epoch - 3ms/step
Model Rated #3 Evaluation:
Loss: 0.550627589225769, Accuracy: 0.7292128205299377
Model Rated #3 Hyperparamaters:


{'activation': 'relu',
 'first_units': 50,
 'num_layers': 3,
 'units_0': 40,
 'units_1': 30,
 'units_2': 20,
 'units_3': 20,
 'units_4': 10,
 'tuner/epochs': 20,
 'tuner/initial_epoch': 0,
 'tuner/bracket': 0,
 'tuner/round': 0}